mod_remove_extra_semicolon removes non-extra semicolons

$ uncrustify --version
Uncrustify_d-811d13a4
$ cat p.cfg
mod_remove_extra_semicolon      = true
$ cat bug.c
#define CONCAT2x(a,b) a ## _ ## b 
#define CONCAT2(a,b) CONCAT2x(a,b)

typedef struct S {
  int a;
  int b;
} S;

static const S CONCAT2(foo, bar) = {
                                   3,
                                   4
};

$ uncrustify -c p.cfg bug.c
Output suffix: .uncrustify
Parsing: bug.c as language C
$ cat bug.c.uncrustify 
#define CONCAT2x(a,b) a ## _ ## b
#define CONCAT2(a,b) CONCAT2x(a,b)

typedef struct S {
	int a;
	int b;
} S;

static const S CONCAT2(foo, bar) = {
	3,
	4
}

$ 

Please note that the pattern shown (use of CONCAT2() or similar macro to construct an identifier) is very much used in code where every module needs identifiers of a certain form (e.g., including an encoding of the module name).

The configuration file p.cfg is a one-liner:
<pre>
mod_remove_extra_semicolon = true
</pre>

The source file p.c is (valid C):
<pre>
#define CONCAT2x(a,b) a ## _ ## b
#define CONCAT2(a,b) CONCAT2x(a,b)

typedef struct S {
int a;
int b;
} S;

static const S CONCAT2(foo, bar) = {
3,
4
};
</pre>
The obtained output (invalid C) is:
<pre>
#define CONCAT2x(a,b) a ## _ ## b
#define CONCAT2(a,b) CONCAT2x(a,b)

typedef struct S {
	int a;
	int b;
} S;

static const S CONCAT2(foo, bar) = {
	3,
	4
}
</pre>
This is invalid because the final semicolon, which is not redundant, has been stripped.
The expected output would be:
<pre>
#define CONCAT2x(a,b) a ## _ ## b
#define CONCAT2(a,b) CONCAT2x(a,b)

typedef struct S {
	int a;
	int b;
} S;

static const S CONCAT2(foo, bar) = {
	3,
	4
};
</pre>

Typescript:
<pre>
$ ./uncrustify --version
Uncrustify_d-0.74.0-225-8edf48ad
$ gcc -c p.c
$ ./uncrustify -c p.cfg -f p.c -o q.c
do_source_file(1507): Parsing: p.c as language C
$ gcc -c q.c
q.c:12:1: error: expected ‘,’ or ‘;’ at end of input
   12 | }
      | ^
1|$ 
</pre>

--- a/src/semicolons.cpp
+++ b/src/semicolons.cpp
@@ -70,7 +70,35 @@ void remove_extra_semicolons(void)
                     || get_chunk_parent_type(prev) == CT_FUNC_CLASS_DEF
                     || get_chunk_parent_type(prev) == CT_NAMESPACE))
          {
-            remove_semicolon(pc);
+            // looking for code block vs. initialisation
+            bool  code_block_found = true;
+            Chunk *closing_brace   = pc->GetPrevNcNnl();                   // Issue #3506
+
+            if (  closing_brace != nullptr
+               && closing_brace->IsNotNullChunk())
+            {
+               Chunk *opening_brace = closing_brace->SkipToMatchRev();
+
+               if (  opening_brace != nullptr
+                  && opening_brace->IsNotNullChunk())
+               {
+                  Chunk *equal_sign = opening_brace->GetPrevNcNnl();
+
+                  if (  equal_sign != nullptr
+                     && equal_sign->IsNotNullChunk()
+                     && chunk_is_token(equal_sign, CT_ASSIGN))
+                  {
+                     // initialisation found
+                     code_block_found = false;
+                  }
+               }
+            }
+
+            if (code_block_found)
+            {
+               // code block found
+               remove_semicolon(pc);
+            }
          }
          else if (  chunk_is_token(prev, CT_BRACE_CLOSE)
                  && get_chunk_parent_type(prev) == CT_NONE)

--- a/tests/c.test
+++ b/tests/c.test
@@ -166,6 +166,7 @@
 00421  c/nl_ds_struct_enum_cmt-t.cfg              c/nl_ds_struct_enum.c
 00422  c/nl_ds_struct_enum-2.cfg                  c/nl_ds_struct_enum.c
 00423  c/bug_1702.cfg                             c/bug_1702.c
+00424  c/Issue_3506.cfg                           c/Issue_3506.c
 
 00430  common/empty.cfg                           c/paren-indent.c
 00431  c/indent_paren_close-1.cfg                 c/paren-indent.c

--- a/tests/config/c/Issue_3506.cfg
+++ b/tests/config/c/Issue_3506.cfg
@@ -1,0 +1,1 @@
+mod_remove_extra_semicolon = true

--- a/tests/expected/c/00424-Issue_3506.c
+++ b/tests/expected/c/00424-Issue_3506.c
@@ -1,0 +1,12 @@
+#define CONCAT2x(a,b) a ## _ ## b
+#define CONCAT2(a,b) CONCAT2x(a,b)
+
+typedef struct S {
+	int a;
+	int b;
+} S;
+
+static const S CONCAT2(foo, bar) = {
+	3,
+	4
+};

--- a/tests/input/c/Issue_3506.c
+++ b/tests/input/c/Issue_3506.c
@@ -1,0 +1,12 @@
+#define CONCAT2x(a,b) a ## _ ## b
+#define CONCAT2(a,b) CONCAT2x(a,b)
+
+typedef struct S {
+int a;
+int b;
+} S;
+
+static const S CONCAT2(foo, bar) = {
+3,
+4
+};

